File: common/src/main/java/cloudcmd/common/engine/ReplicationStrategy.java
package cloudcmd.common.engine;

import cloudcmd.common.adapters.Adapter;

import java.io.InputStream;
import java.util.Set;

public interface ReplicationStrategy {
  public boolean isReplicated(Set<Adapter> adapters, String hash) throws Exception;
  public void push(CloudEngineListener listener, Set<Adapter> adapters, String hash) throws Exception;
  public InputStream load(String hash) throws Exception;
}

File: common/src/main/scala/cloudcmd/common/engine/ReplicationStrategy.scala
package cloudcmd.common.engine

import cloudcmd.common.adapters.DirectAdapter
import java.io.InputStream
import cloudcmd.common.BlockContext

trait ReplicationStrategy {
  def isReplicated(ctx: BlockContext, adapters: List[DirectAdapter]): Boolean

  def store(ctx: BlockContext, is: InputStream, adapters: List[DirectAdapter])

  def load(ctx: BlockContext, adapters: List[DirectAdapter]): (InputStream, Int)

  def remove(ctx: BlockContext, adapters: List[DirectAdapter]) : Boolean

  def ensure(ctx: BlockContext, hashProviders: List[DirectAdapter], adapters: List[DirectAdapter], blockLevelCheck: Boolean) : Boolean
}